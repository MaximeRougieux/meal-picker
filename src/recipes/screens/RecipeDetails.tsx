import React from 'react';
import { ScrollView, Text, Image, View, StyleSheet } from 'react-native';
import { RecipeDetailsProps } from 'utils/navigation/types';
import { Meal } from 'recipes/types/APIMealsResponseData';

interface RecipeInstructionProps {
  item: string;
  index: number;
}

interface InstructionsProps {
  meal: Meal;
}

const RecipeInstruction = ({ item, index }: RecipeInstructionProps) => {
  return (
    <View style={styles.instructionContainer}>
      <Text>{`${index + 1}.   `}</Text>
      <Text>{item}</Text>
    </View>
  );
};

const Instructions = ({ meal }: InstructionsProps) => {
  const textInstructions = meal.strInstructions
    .split('\r\n')
    .filter(instruction => instruction);

  return (
    <>
      {textInstructions.map((item, index) => (
        <RecipeInstruction item={item} index={index} key={index} />
      ))}
    </>
  );
};
export const RecipeDetails = ({ route }: RecipeDetailsProps) => {
  const { meal } = route.params;
  return (
    <ScrollView>
      <Image source={{ uri: meal.strMealThumb }} style={styles.image} />
      <View style={styles.container}>
        <Text style={styles.title}>{meal.strMeal}</Text>
        <Text style={styles.category}>{meal.strCategory}</Text>
        <Instructions meal={meal} />
      </View>
    </ScrollView>
  );
};

const styles = StyleSheet.create({
  image: {
    height: 300,
    width: '100%',
    borderBottomLeftRadius: 50,
    borderBottomRightRadius: 50,
  },
  category: {
    alignSelf: 'flex-end',
    fontSize: 20,
    fontStyle: 'italic',
  },
  container: {
    marginHorizontal: 12,
    marginTop: 16,
  },
  title: {
    color: 'black',
    fontSize: 36,
    fontFamily: 'Helvetica',
    fontWeight: '500',
    alignSelf: 'center',
    textAlign: 'center',
    marginBottom: 24,
  },
  instructionContainer: {
    flexDirection: 'row',
    margin: 8,
  },
});
