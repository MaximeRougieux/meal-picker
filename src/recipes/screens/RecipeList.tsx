import React, { useState } from 'react';
import {
  ActivityIndicator,
  ScrollView,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import { RecipeListProps } from 'utils/navigation/types';
import { RecipeItem } from '../components/RecipeItem';
import SearchBar from '../components/SearchBar';
import useGetMeals from '../usecases/useGetMeals';

export const RecipeList = ({ navigation }: RecipeListProps) => {
  const [keyword, setKeyword] = useState('');
  const { loading, error, items } = useGetMeals(keyword);

  return (
    <View style={styles.main}>
      <Text style={styles.title}>Welcome to Meal Picker</Text>
      <SearchBar setKeyword={setKeyword} />
      <ScrollView>
        {loading ? (
          <ActivityIndicator testID="loading" />
        ) : error !== null ? (
          <Text style={styles.errorText}>{error}</Text>
        ) : items.length === 0 ? (
          <Text style={styles.errorText}>No meal available</Text>
        ) : (
          items.map(meal => (
            <TouchableOpacity
              key={meal.idMeal}
              onPress={() => navigation.navigate('Details', { meal: meal })}>
              <RecipeItem
                name={meal.strMeal}
                category={meal.strCategory}
                imageUrl={meal.strMealThumb}
              />
            </TouchableOpacity>
          ))
        )}
      </ScrollView>
    </View>
  );
};

const styles = StyleSheet.create({
  main: {
    margin: 8,
  },
  errorText: {
    fontSize: 16,
    alignSelf: 'center',
  },
  title: {
    marginVertical: 16,
    fontSize: 36,
  },
});
