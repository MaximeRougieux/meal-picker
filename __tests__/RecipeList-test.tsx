import React from 'react';
import nock from 'nock';
import { RecipeList } from 'recipes/screens/RecipeList';
import {
  render,
  waitForElementToBeRemoved,
} from '@testing-library/react-native';

test('renders correctly', async () => {
  nock('https://www.themealdb.com/api/json/v1/1')
    .get('/search.php?s=')
    .reply(200, {
      meals: [
        {
          idMeal: '52771',
          strMeal: 'Spicy Arrabiata Penne',
          strCategory: 'Vegetarian',
          strMealThumb:
            'https://www.themealdb.com/images/media/meals/ustsqw1468250014.jpg',
        },
      ],
    });
  const route: any = {};
  const navigation: any = {};
  const { getByTestId, toJSON, getByText } = render(
    <RecipeList route={route} navigation={navigation} />,
  );
  await waitForElementToBeRemoved(() => getByTestId('loading'));
  expect(getByText('Spicy Arrabiata Penne')).toBeTruthy();
  expect(toJSON()).toMatchSnapshot();
});
