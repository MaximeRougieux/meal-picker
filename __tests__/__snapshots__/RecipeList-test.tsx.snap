// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <TextInput
    onChangeText={[Function]}
    placeholder="Search"
    style={
      Object {
        "fontSize": 18,
        "margin": 6,
      }
    }
    testID="searchbar"
  />,
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 200,
            "margin": 8,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "https://www.themealdb.com/images/media/meals/ustsqw1468250014.jpg",
            }
          }
          style={
            Object {
              "aspectRatio": 1,
              "height": "100%",
              "marginRight": 8,
            }
          }
        />
        <View
          style={
            Object {
              "flex": 1,
              "height": "100%",
              "justifyContent": "flex-start",
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 24,
              }
            }
          >
            Spicy Arrabiata Penne
          </Text>
          <Text
            style={
              Object {
                "fontSize": 12,
              }
            }
          >
            Vegetarian
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>,
]
`;
